### ISSUE TITLE
Allow per-buffer setup for each language server.

### ISSUE THREAD

**Issue by no-more-secrets** [neovim/nvim-lspconfig#2627]
### Language server

lua-language-server

### Requested feature

In my source tree have some Lua files that require one version of Lua, and others that require another.  But when I configure lua-language-server via nvim-lspconfig I can only call `setup` once, whereas I would like to be able to configure the same language server differently depending on the file path of the file being opened.  Ideally I'd have a scenario where I can have nvim-lspconfig launch two instances of the language server, one handling all of the Lua 5.1 files and the other handling all of the Lua 5.4 files.

Any way to do this and/or add this feature?

**Comment by glepnir** [comment]
lua_ls support lua5.1 - lua 5.4 check lua_ls wiki for `runtime`

**Comment by no-more-secrets** [comment]
@glepnir I am aware that lua_ls supports multiple language versions, but the version can only be specified during the `setup` call.  However, nvim-lspconfig does not seem to allow per-buffer setup calls, so I cannot set a separate Lua version per buffer, and that is my feature request.  I think you might have misunderstood the question -- can we reopen?

**Comment by gwerbin** [comment]
@no-more-secrets I haven't looked at the Lua config in detail, but I think in the default configuration, nvim-lspconfig will start a separate lang server instance for each root directory. You might be able to change the Lua version inside the `on_new_config` callback or similar.

Just based on my reading of the docs, you should be able to do something like this:

```lua
if vim.fn.executable("lua-ls-wrapper.sh") ~= 0 then
  lspconfig.lua_ls.setup({

    on_new_config = function(new_config, root_dir)
      local lua_version = "Lua 5.4"
      if ____ then  -- Whatever condition you want here
        lua_version = "LuaJIT"
      end

      new_config.settings = {
        Lua = { runtime = { version = lua_version } },
      }
    end

  })
end
```

For example, you can put a `.lua-version` file in the root directory of the project, and you can look for that file in `root_dir` and read its contents to get the required project version. The point is that you can do whatever you want and apply whatever settings you want by modifying `new_config.settings` in the `on_new_config` callback.

**Comment by no-more-secrets** [comment]
@gwerbin Thanks; what I ended doing was just dropping some `.luarc.json` files in each root folder which contain the Lua version (and other things as well), and the nvim-lspconfig for lua-language-server is already configured to look for those to define root folders.  The lua-language-server seems also to search for and respect those files and allows them to override what is specified in that one-time configuration in the Lua code.  Overall, that seems to work for me (modulo one bug that I raised in the lua-language-server repo).

FYI, something that I observed is that I can open a single editor session containing multiple Lua files of different versions (specified in the .luarc.json files) but there is only one instance (process) of lua-language-server running, which I guess is even better if it supports that, which apparently it does.
